tests/ui/lint/lint-ctypes-cstr.rs

```diff
@@ -0,0 +1,26 @@
+#![crate_type = "lib"]
+#![deny(improper_ctypes, improper_ctypes_definitions)]
+
+use std::ffi::{CStr, CString};
+
+extern "C" {
+    fn take_cstr(s: CStr);
+    //~^ ERROR `extern` block uses type `CStr`, which is not FFI-safe
+    //~| HELP consider using `*const std::ffi::c_char` instead, and pass the value from `CStr::as_ptr()`
+    fn take_cstr_ref(s: &CStr);
+    //~^ ERROR `extern` block uses type `CStr`, which is not FFI-safe
+    //~| HELP consider using `*const std::ffi::c_char` instead, and pass the value from `CStr::as_ptr()`
+    fn take_cstring(s: CString);
+    //~^ ERROR `extern` block uses type `CString`, which is not FFI-safe
+    //~| HELP consider using `*const std::ffi::c_char` instead, and pass the value from `CStr::as_ptr()`
+    fn take_cstring_ref(s: &CString);
+    //~^ ERROR `extern` block uses type `CString`, which is not FFI-safe
+    //~| HELP consider using `*const std::ffi::c_char` instead, and pass the value from `CStr::as_ptr()`
+}
+
+extern "C" fn rust_take_cstr_ref(s: &CStr) {}
+//~^ ERROR `extern` fn uses type `CStr`, which is not FFI-safe
+//~| HELP consider using `*const std::ffi::c_char` instead, and pass the value from `CStr::as_ptr()`
+extern "C" fn rust_take_cstring(s: CString) {}
+//~^ ERROR `extern` fn uses type `CString`, which is not FFI-safe
+//~| HELP consider using `*const std::ffi::c_char` instead, and pass the value from `CStr::as_ptr()`
```

Comment:
Should we adapt the message for mutable references/pointers?

Reply:
The lint already triggers for mutable references/pointers. Do you mean change what the error message says instead? What do you think it should say?

Reply:
@cjgillot Any update?

Reply:
Yes, I mean adapting the message. When we expect a `&mut CString`, should we suggest passing `CStr::as_mut_ptr()` instead?

Reply:
@cjgillot I've rephrased the lint message to be independent of mutability - does the current message seem acceptable to you? If not, I can have the message change conditionally, but:
- It would require a more significant code refactor (`ImproperCTypesVisitor::check_type_for_ffi()` doesn't keep track of the base type before recursing)
- AFAIK the only valid way to obtain a mutable pointer is to own a `CString` and call `.into_raw()`, but this requires `CString::from_raw()` to be called at a later point and be dropped. I think if we leave the lint message vague enough it will lead the user to reassess. `CString::into_raw()` isn't a perfect analogue to `&mut CString` anyways. And `CStr` can never be made mutable because it can be aliased (CString Derefs to it).

What do you think?

Reply:
r=me with a negative test, to verify that we do not suggest anything on `&mut CString`.

Reply:
@cjgillot Done.